### DISCUSSION TITLE
invalid_client OAuthCallback error for custom oauth provider (mastodon)

### DISCUSSION THREAD

**Discussion by jacksonopp** [nextauthjs/next-auth#5801] [Help]
## Brief Description of issue
I'm trying to create an application that allows me to log in with a mastodon account. I've created a custom Mastodon provider (copied from [this currently open pr](https://github.com/nextauthjs/next-auth/pull/5739) #5739  and just set up in util.ts file).

I'm able to get to the mastodon authorization page, but after authorizing, i'm redirected to the `/api/auth/signin?error=OAuthCallback` route. The server error I'm getting is:

```
{
    name: 'OAuthCallbackError',
    code: undefined
  },
  providerId: 'mastodon',
  message: 'invalid_client (Client authentication failed due to unknown client, no client authentication included, or unsupported authentication method.)'
}
```

## What i've done so far
I have properly (i think) set up an application on my mastodon server with the proper client id and secret as well as the following redirect uri: `http://localhost:3000/api/auth/callback/mastodon`

My providers in my `[...nextAuth].ts` file are as follows:

```ts
// importing the custom mastodon provider function
import Mastodon from './path/to/custom/provider/file'

export const authOptions: NextAuthOptions = {
  callbacks: {
    session({ session, user }) {
      if (session.user) {
        session.user.id = user.id;
      }
      return session;
    },
  },
  providers: [
    Mastodon({
      issuer: process.env.MASTODON_ISSUER,
      clientId: process.env.MASTODON_CLIENT_ID,
      clientSecret: env.MASTODON_ISSUER,
    })
  ]
}
```
And my login button is simply:

```tsx
<button onClick={() => signIn()}>
```

## My config
next: 13.0.2
next-auth: 4.15.1

OS: MacOS Ventura

## Screenshots
<figure>
<img width="580" alt="Screenshot 2022-11-11 at 10 06 38 PM" src="https://user-images.githubusercontent.com/50721527/201456412-4a5a2481-5a22-4a0a-9f89-3f721111bb81.png">
<figcaption>This is what the mastodon auth page looks like. Clicking the "authorize" button takes me to the error page and throws the error in the server logs.</figcaption>
</figure>

**Comment by eiiot** [comment]
Did you ever manage to solve this? I'm running into a similar issue using a different custom provider

**Reply by abdullahahmeda** [reply]
Same issue here. I also created a discussion here: https://github.com/nextauthjs/next-auth/discussions/10204

**Reply by eiiot** [reply]
I haven't worked on the project in a while, but I would look into this Stack Overflow thread:

https://stackoverflow.com/questions/76399264/nextauth-gives-error-oauth-callback-error-invalid-client-when-authenticating-wit

**Reply by abdullahahmeda** [reply]
I have just fixed it by adding:
```ts
client: {
  token_endpoint_auth_method: 'client_secret_post'
}
```
More info can be found [here](https://developer.okta.com/docs/api/openapi/okta-oauth/guides/client-auth/).

I appreciate your help though. Thank you!

**Reply by abdullahahmeda** [reply]
I didn't notice that the link you provided uses the same code 😃
